pilot/info/storageactivitymaps.py

```diff
@@ -11,20 +11,34 @@
 logger = logging.getLogger(__name__)
 
 
-sched2ddm_activity_map = {'es_events': 'pw',
-                          'es_failover': 'pw',
-                          'es_events_read': 'pr'}
+sched2ddm_activity_map_lan = {'es_events': 'write_lan',
+                              'es_failover': 'write_lan',
+                              'es_events_read': 'read_lan'}
 
 
-def get_ddm_activity(activity):
+sched2ddm_activity_map_wan = {'es_events': 'write_wan',
+                              'es_failover': 'write_wan',
+                              'es_events_read': 'read_wan'}
+
+
+def get_ddm_activity(activity, queuedata, ddmendpoint, ddm):
     """
     Map schedconf activities to corresponding ddm activities
 
     :param activity: schedconf activity as string
+    :param queuedata: Queue data info.
+    :param ddmendpoint: ddm endpoint.
+    :param ddm: ddm storage data.
     :return: ddm activity as string
     """
-    if activity in sched2ddm_activity_map:
-        ddm_activity = sched2ddm_activity_map[activity]
+    if activity in sched2ddm_activity_map_lan:
```

Comment:
Wen, the translation  looks a bit confusing since it mixed together activity mapping and the decision which access mode to use (WAN or LAN), so that as result it cross depends on other logic above and requires extra redundant data to be passed.

What is original logic you used to translate activity name?

 As I see for ES-workflow you always pass `queuedata=None`, so that all this logic implemented  here has no effect and for ES is equivalent to one 1 line call 

`ddm_activity = sched2ddm_activity_map_wan.get(activity, activity)`

May be it's better to apply data look up at upper level? but here just take only map(s) `sched2ddm_activity_map_lan ` or `sched2ddm_activity_map_wan `

Reply:
1) As I see for ES-workflow you always pass queuedata=None:
Not always.
get_ddm_activity is called in two places. One is in pilot/api/data.py. It's called with correct queuedata.
Second is called in objectstore.py. It's fixed with queuedata now.

2) Here is the idea:
At first, try to map es_* activities to lan activities. But if the output rse is not local, it will cause problems. So for example, if it maps es_events to write_lan, it needs to check whether the ddmendpoints in the astorages['write_lan']. If the ddmendpoint is not in the astorages['write_lan'], it means the ddmendpoint is not a local storage. So I should use 'write_wan'.

Reply:
OK, now I see your logic.

I would suggest to implement this in more flexible way, so that  by default from data API  we get clean workflow to use LAN or WAN protocols for requested ddm for individual file. 

Actually, we don't need at all the ddm activity mapping. We just need to explicitly specify the list of ddm activity names (e.g. ['write_lan', 'write_wan'] by default for stage-out mode) at the top and save in into `FileSpec`, using new field  for example `fspec.ddm_activity`. StageClient by default (if it's not already set in passed `files`) should populate in `transfer` method for each file the `fspec.ddm_activity` list.

Since `FileSpec` instance already knows the list of local ddms (`fspec.inputddms`, it becomes populated in `resolve_replicas` function, this part needs to be reallocated into `resolve_imputddms` and executed in `transfer()`), and in data API while doing `transfer()`  we know the transfer mode (`stage-in` or `stage-out`) as well, so that it will be enough just to prepare by default the list of ddm_activity for given mode for given `fspec` taking into account if file is local (fspec.ddmendpoint in fspec.inputddms)

e.g. something like this inside the `transfer`:
```
for fspec in files:
  if fspec.ddm_activity:  # skip already initialized data
      continue
  if self.mode == 'stage-in'
     fspec.ddm_activity = filter(None, ['read_lan' if fspec.ddm in fspec.inputddms else None, 'read_wan']
  else:
     fspec.ddm_activity = filter(None, ['write_lan' if fspec.ddm in fspec.inputddms else None, 'write_wan']
```
and then remove `activity` as input for `resolve_protocols` function, but rather consider individual settings per file (fspec.ddm_activity).  

This implementation allows to control LAN and WAN settings per file. e.g. to fetch LIB files over WAN, while copy2scratch root file via LAN protocols.


Wen, can you do this? I can push a MR if your prefer.

thanks

Reply:
it's good that you do it.

Reply:
OK, sounds good, will submit soon

Reply:
I refactored ES Stage clients in https://github.com/PanDAWMS/pilot2/pull/185 , the updates in particular covers LAN and WAN protocol preference for file transfer.